### PR TITLE
UX-600/Created Terraform Tab and API Helper Page and Fixed DevTools ApiHelper

### PR DESCRIPTION
**Pull request by knnguy**
**Exposed API Services**

- Qbert
- Keystone
- Appbert (only if dev plugin is enabled)
- Resmgr (only if dev plugin is enabled)

**DevTools ApiHelper**

- Works now (Couldn't make requests before)

**API Details Page**

- Search bar allows user to search by method name
- User can also filter by request verb (PUT, PATCH, GET, POST, DELETE)

**API Request Helper**

- ClusterUuid and CloudProviderUuid params have a dropdown menu so users can select an options vs. having to type it out
- The URL field will dynamically show the the url with the parameters that the user enters
- The Make API Request button and the JSON response CodeMirror will only show up on GET methods which means that making PUT, PATCH, DELETE, and POST requests will be disabled for users. However, if dev plugin is enabled, then everything is allowed

### **Added tabs and a link to the API Helper in the API Access Page**
Screen Shot 2020-10-14 at 6 37 05 PM

### **New Terraform Tab**
Screen Shot 2020-10-14 at 6 37 22 PM

### **API Helper Page**
Screen Shot 2020-10-14 at 6 39 37 PM

### **Error Message Pop Up When Request Fails**
Screen Shot 2020-10-14 at 4 03 26 PM

### **Make API Request button and CodeMirror will not show up on non GET methods unless ui-dev is enabled**
Screen Shot 2020-10-14 at 8 48 34 PM

### CODE REVIEW

**Review by jruddell — CHANGES_REQUESTED** [review]
Great work! This is going to be really useful! I left some optimization / change comments down to address 😄

File: src/app/plugins/developer/components/ApiHelper.js [hunk 1]
```diff
@@ -69,8 +79,9 @@ class ApiHelper extends React.PureComponent {
     this.setState({ service })
 
     if (service === 'qbert') {
-      return this.setState({ baseUrl: await apiClient.qbert.baseUrl() })
+      return this.setState({ baseUrl: await apiClient.qbert.getApiEndpoint() })
```
[the remark this answers is not in the document]
Comment (knnguy) [reply src/app/plugins/developer/components/ApiHelper.js:82]: I don't see the endpoint being stored in Qbert

Screen Shot 2020-10-15 at 10 49 34 AM

File: src/app/plugins/kubernetes/components/apiAccess/ApiRequestHelper.tsx [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+  const devEnabled = window.localStorage.enableDevPlugin === 'true'
+  const { params, getParamsUpdater } = useParams()
+  const [apiResponse, setApiResponse] = React.useState('')
+  const [requestUrl, setRequestUrl] = React.useState('')
+  const apiClient = ApiClient.getInstance()
+  const showToast = useToast()
+
+  useEffect(() => {
+    setApiResponse('')
+    ;(async () => {
```
[the remark this answers is not in the document]
Comment (knnguy) [reply src/app/plugins/kubernetes/components/apiAccess/ApiRequestHelper.tsx:137]: I actually copied that syntax from the useEffect hook in  `SummaryStep.tsx`. I'll go ahead and change in that file as well.

File: src/app/plugins/developer/components/ApiHelper.js [hunk 3]
```diff
@@ -48,15 +48,25 @@ class ApiHelper extends React.PureComponent {
...
-      GET: () => apiClient.basicGet(finalUrl),
-      POST: () => apiClient.basicPost(finalUrl, JSON.parse(body)),
-      PUT: () => apiClient.basicPut(finalUrl, JSON.parse(body)),
-      DELETE: () => apiClient.basicGet(finalUrl),
+      GET: () => apiClient.basicGet({ url, endpoint: baseUrl, options: { clsName: service } }),
+      POST: () =>
+        apiClient.basicPost({
+          url,
+          endpoint: baseUrl,
+          body: JSON.parse(body),
```
[the remark this answers is not in the document]
Comment (knnguy) [reply src/app/plugins/developer/components/ApiHelper.js:59]: User enters the request body in a TextField so the value is a string. That's why we need to turn it into an object

File: src/app/plugins/kubernetes/components/apiAccess/TerraformListPage.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import React from 'react'
```
Comment (jruddell) [inline src/app/plugins/kubernetes/components/apiAccess/TerraformListPage.js:1]: All new files should be typescript (i'm addressing this currently)

File: src/app/plugins/kubernetes/components/apiAccess/TerraformListPage.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    marginTop: theme.spacing(6),
+  },
+  buttonContainer: {
+    marginTop: theme.spacing(4),
+  },
+}))
+
+const requirements = ['Go v1.13+ and', 'Terraform']
+
+const iconSizes = { small: '', medium: '@2x', large: '@3x' }
```
[the remark this answers is not in the document]
Comment (knnguy) [reply src/app/plugins/kubernetes/components/apiAccess/TerraformListPage.js:53]: I did change it use CloudProviderCard, but every time I hover over it, it has a border. Doesn't seem like that can be disabled in CloudProviderCard. Should I go ahead and modify CloudProviderCard?

Screen Shot 2020-10-15 at 5 49 38 PM

File: src/app/plugins/kubernetes/components/apiAccess/ApiRequestHelper.tsx [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  },
+  parameters: {
+    display: 'flex',
+    flexFlow: 'column nowrap',
+    marginBottom: theme.spacing(2),
+  },
+}))
+
+const renderParamField = ({ classes, paramName, params, getParamsUpdater }) => {
+  if (paramName === 'clusterUuid') {
```
[the remark this answers is not in the document]
Comment (knnguy) [reply src/app/plugins/kubernetes/components/apiAccess/ApiRequestHelper.tsx:74]: Not sure how you want me to do this. Map `paramName` to a component?

File: src/app/plugins/kubernetes/components/apiAccess/ApiRequestHelper.tsx [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        POST: () => apiClient.basicPost(params),
+        PATCH: () => apiClient.basicPatch(params),
+        PUT: () => apiClient.basicPut(params),
+        DELETE: () => apiClient.basicDelete(params),
+      }[metadata.type]()
+
+      return response
+    } catch (e) {
+      typeof e === 'object'
+        ? showToast(e.message, MessageTypes.error)
```
[the remark this answers is not in the document]
Comment (knnguy) [reply src/app/plugins/kubernetes/components/apiAccess/ApiRequestHelper.tsx:168]: Also not sure what you mean by this? Is there a component here I'm supposed to use to show the error message?

**Pull request closed without merging**